gravitational/teleport
# Add Support for Oracle protocol

## What
Add Oracle Protocol Integration

## Changes: 
 - [RFD](https://github.com/gravitational/teleport/pull/16435) 
 - tsh local proxy listener on TLS port with local self signed certs where oracle client uses a wallet with local proxy certs.
 - `tctl auth sign --format=oracle` command detects if Orapki binary is avable in user env in order to create Oracle wallet. Otherwise a user need manual steps to convert teleport certs to oracle wallet format. 
 -

## UX:
- tctl example:
   ```
    $ tctl  auth sign --format=oracle --host=localhost --out=certs/server --ttl=2190h
    To enable mutual TLS on your Oracle server, add the following settings to oracle sqlnet.ora configuration file:
    WALLET_LOCATION = (SOURCE = (METHOD = FILE)(METHOD_DATA = (DIRECTORY = /path/to/oracleWalletDir)))
    SSL_CLIENT_AUTHENTICATION = TRUE
    SQLNET.AUTHENTICATION_SERVICES = (TCPS)
    To enable mutual TLS on your Oracle server, add the following TCPS entreis to its listener.ora configuration file:
    LISTENER =
      (DESCRIPTION_LIST =
        (DESCRIPTION =
          (ADDRESS = (PROTOCOL = TCPS)(HOST = 0.0.0.0)(PORT = 2484))
        )
      )
    WALLET_LOCATION = (SOURCE = (METHOD = FILE)(METHOD_DATA = (DIRECTORY = /path/to/oracleWalletDir)))
    SSL_CLIENT_AUTHENTICATION = TRUE
    ```
- tsh proxy db: 
  ```
  $ tsh proxy db oracle  --db-user=alice --db-name=XE --tunnel
      Started authenticated tunnel for the Oracle database "oracle" in cluster "ice-berg.dev" on 127.0.0.1:51584.
      To avoid port randomization, you can choose the listening port using the --port flag.
      Use the following command to connect to the Oracle database server using CLI:
        $ sql -L jdbc:oracle:thin:@tcps://localhost:51584/XE?TNS_ADMIN=/Users/marek/.tsh/keys/ice-berg.dev/marek-db/ice-berg.dev/oracle-wallet
      or using following Oracle JDBC connection string in order to connect with other GUI/CLI clients:
        jdbc:oracle:thin:@tcps://localhost:51584/XE?TNS_ADMIN=/Users/marek/.tsh/keys/ice-berg.dev/marek-db/ice-berg.dev/oracle-wallet
  ```
## Testing: 
1) checkout `e` related changes: `smallinsky/oracle_protocol_v3` https://github.com/gravitational/teleport.e/pull/1011
2) build Teleprot Enterprise 
3) Setup local Oracle Database https://github.com/smallinsky/infra/tree/main/oracle
   - `cd infra/oracle`
   - `make` make sure that you are current logged into teleport cluster with a user that has a permission to issue DB certs.



## Next Steps: 
- Add support for audit logs
- Align Teleport Connect
